### ISSUE TITLE
JMS AutoConfig & JMSListener config enrichment

### ISSUE THREAD

**Issue by gokhanoner** [spring-projects/spring-boot#3390]
`sessionAcknowledgeModeName` and `sessionTransacted` options, and some others necessary when using local transactions, but auto-configured container just use `pubSubDomain` option. It'd be good to define these parameters through `spring.jms.` properties on `JmsAnnotationDrivenConfiguration`  or maybe these can be added to the `JMSListener` annotation.

**Comment by snicoll** [comment]
I think we should be able to auto-detect the `sessionTransacted` flag. Regarding acknwoledgment mode and other settings, I agree that it might be nice to expose properties for them. I am not sure where to draw a line though.

**Comment by gokhanoner** [comment]
Well, concurrency and subscription can be set on JMSListener, so I suppose session options are only simple options, others require Handlers o Resolvers so it's not possible, I guess, to define them through application.properties.

**Comment by snicoll** [comment]
Basically checking if a suitable transaction manager is present and if not flip the `sessionTransacted` flag to true. We already check for the transaction manager use case so it's a matter of adding an else block.

**Comment by gokhanoner** [comment]
But in the transaction manager case, there is a thing. In my case, I'm also using JPA transactions, but I cannot use XA so, in ```JmsAnnotationDrivenConfiguration```, ```JtaTransactionManager``` is not null and points my JPA transaction manager? That means that I won't be configured as expected, right ?

**Comment by snicoll** [comment]
It does not anymore (#3150). It's a bug and it was fixed. Using the `JpaTransactionManager` for the JMS container is plain wrong actually.

**Comment by gokhanoner** [comment]
Yep, just see the commit. I guess I need to write my own ```DefaultJmsListenerContainerFactory``` bean for the time being, until 1.3.0 is out.
Thanks Stephane.

**Comment by snicoll** [comment]
or not ... Check #3393 (`1.2.5` should be released this week). Thanks for the feedback and let's keep this issue open for the other customizations you were mentioning.

**Comment by gokhanoner** [comment]
yes, just see it and it's already closed :+1:  Fantastic.

**Comment by gokhanoner** [comment]
@snicoll Regaring #3393 , I did a detailed test today, and actually, current config, PlatformTransactionManager, works as documented. Actually, it sync the JPA transaciton and JMS transacitons and if one fails, the other also.

```
	@JmsListener(destination = DirectorySwiftSender.SWIFT_MESSAGE_QUEUE)
	public void convertFileToSwiftMessage(Message<String> message) {
		log.info("{}", message.getHeaders());
		storageService.logMessageRead(message.getHeaders().get(DirectorySwiftSender.FILE_NAME_PROP, String.class), 
				message.getHeaders().get(DirectorySwiftSender.FILE_ORDER_PROP, Integer.class));
		if("1".equals(message.getHeaders().get("JMSXDeliveryCount", String.class))) {
			throw new RuntimeException("ilkinde hata ver");
		}
		log.info("message processed");
	}
```

```storageService.logMessageRead() ``` annotated as ```@Transacted``` and it workd as expected in 1.2.4: db update occur only once. But when I apply your new changes, it doesn't apply any sessionTransaction and for above code, I see 2 records on the DB.

Basically, the current behavior is correct: if there is a Transaction manager, JPA or JTA, it align itself automatically and session became a transacted session.

What do you think?

**Comment by snicoll** [comment]
This doesn't seem right. A `JpaTransactionManager` is **not** meant to synchronize the JMS session. Are you saying that `1.2.5.BUILD-SNAPSHOT` is broken?

**Comment by gokhanoner** [comment]
No, I'm saying that 1.2.4 working correct according to this: ```AbstractMessageListenerContainer#setSessionTransacted```
```
Set the transaction mode that is used when creating a JMS Session. Default is "false".
Note that within a JTA transaction, the parameters passed to create(Queue/Topic)
Session(boolean transacted, int acknowledgeMode) method are not taken into account.
Depending on the J2EE transaction context, the container makes its own decisions on these values.
Analogously, these parameters are not taken into account within a locally managed transaction
either, since the accessor operates on an existing JMS Session in this case. 

Setting this flag to "true" will use a short local JMS transaction when running outside of a 
managed transaction, and a synchronized local JMS transaction in case of a managed transaction
 (other than an XA transaction) being present. This has the effect of a local JMS transaction being
 managed alongside the main transaction (which might be a native JDBC transaction), with the JMS 
transaction committing right after the main transaction.

Parameters:sessionTransacted See Also:javax.jms.Connection.createSession(boolean, int)
```

So according to this, current behavior is correct. But after #3150 change, this feature is basically lost.

**Comment by snicoll** [comment]
I disagree;

1. #3150 prevents to inject something else than a `JtaTransactionManager` in the listener container. It makes no sense to associate, say a `JpaTransactionManager` to the JMs infra. So that change is necessary
2. If no transaction manager is set at all (because there 's no JTA infra available per 1), you need to flip the `sessionTransacted` flag to `true` (per the documentation that you just provided yourself).

It's still unclear  to me and we're having this discussion on the wrong issue anyway. If you have a sample that shows things are broken, please create an issue or associate it on the right one. Thanks ;)

**Comment by snicoll** [comment]
On a second thought, I don't think we should do this for now. `sessionTransacted` is auto-detected so having a flag to set it would just be confusing. The other options can be set with a very simple bean definition and I don't think it's wise to expose all the settings of the default listener container in configuration keys.
